includes/class-wc-stripe-intent-controller.php

```diff
@@ -588,6 +590,61 @@ public function update_order_status_ajax() {
 		}
 	}
 
+	/**
+	 * Handle AJAX request if error occurs while confirming intent.
+	 * We will log the error and update the order.
+	 *
+	 * @throws WC_Stripe_Exception
+	 */
+	public function update_failed_order_ajax() {
+		try {
+			$is_nonce_valid = check_ajax_referer( 'wc_stripe_update_failed_order_nonce', false, false );
+			if ( ! $is_nonce_valid ) {
+				throw new WC_Stripe_Exception( 'missing-nonce', __( 'CSRF verification failed.', 'woocommerce-gateway-stripe' ) );
+			}
+
+			$order_id  = isset( $_POST['order_id'] ) ? absint( $_POST['order_id'] ) : null;
+			$intent_id = isset( $_POST['intent_id'] ) ? wc_clean( wp_unslash( $_POST['intent_id'] ) ) : '';
+			$order     = wc_get_order( $order_id );
+
+			if ( ! empty( $order_id ) && ! empty( $intent_id ) && is_object( $order ) ) {
+				$payment_needed = 0 < $order->get_total();
+				if ( $payment_needed ) {
+					$intent = WC_Stripe_API::retrieve( "payment_intents/$intent_id" );
+				} else {
+					$intent = WC_Stripe_API::retrieve( "setup_intents/$intent_id" );
+				}
+				$error = $intent->last_payment_error;
+
+				if ( ! empty( $error ) ) {
+					WC_Stripe_Logger::log( 'Error when processing payment: ' . $error->message );
+					throw new WC_Stripe_Exception( __( "We're not able to process this payment. Please try again later.", 'woocommerce-gateway-stripe' ) );
+				}
+
+				// Use the last charge within the intent to proceed.
+				if ( isset( $intent->charges ) ) {
+					$charge = end( $intent->charges->data );
+					$this->process_response( $charge, $order );
+				} else {
+					// TODO: Add implementation for setup intents.
+					$this->process_response( $intent, $order );
+				}
```

Comment:
In most cases, we should discover an error on the intent inside `last_payment_error`, but just in case the error is clean somehow, but there is an error or something else unusual on the charge, `process_response` can handle this case: if the status on the charge is `failed`, `process_response` will throw an exception and we will fail the order.